### ISSUE TITLE
"Get rsv1: 0, rsv2: 1, rsv3: 0, they all must be equal to 0" with large arrays sent

### ISSUE THREAD

**Issue by EtienneWan** [hoaproject/Websocket#46]
I'm trying to send big arrays filled with floats to Hoa\WebSocket\Server.
I wrote a client for testing purposes : https://squadclient-html5-etiennewan-1.c9.io/test.html

There is no problem with little arrays (with a length of 1, 10 or an hundred items)
When sending bigger arrays, the WebSocket connection closes unexpectedly without errors server-side and with this reason sent to the client : 
`"Get rsv1: 0, rsv2: 1, rsv3: 0, they all must be equal to 0"` with one or more rsv equal to 1.

Here is the PHP server code (as simple as codes in examples):
```php
<?php
require_once 'vendor/autoload.php';

$server = new Hoa\Websocket\Server(
    new Hoa\Socket\Server('tcp://'.$argv[1].':'.$argv[2])
);

$server->on('message', function (Hoa\Core\Event\Bucket $bucket) {
    $data = $bucket->getData();

    $received = count(explode(",", $data['message']))-1;

    echo $received, PHP_EOL;

    $source = $bucket->getSource();

    $source->send($received);

    return;
});

$server ->on('open', function (){
    echo 'new client connection', PHP_EOL;

});

echo "Server is starting", PHP_EOL;
$server->run();
```

Note : the client and server are hosted on Cloud9 and will be running for debugging purposes.
A live-debugging sessions is possible on the server workspace. Please tell me if you want to have access to it.

Addendum: The size of the array needed to reproduce the error can vary, but 800 seems to be the minimum sometimes.
Here, I only work with normal arrays, but it seemed to happened more often with typed arrays (I was trying to send Float32Array (1024 items) when the error occured for the first time)

**Comment by EtienneWan** [comment]
Modified the "echo $data['message']" line, it is more interesting to count the number of items the server received and return it to the client. Updated in the first message.
When the server close the connection unexpectedly, I can see that a part of the data has been received (for example 942 of 1024 floats).

**Comment by Hywan** [comment]
@EtienneWan With what browser are you testing?

**Comment by Hywan** [comment]
@EtienneWan I just tested your link with Firefox (nightly) with payload of size 256, 800, 1000 and 10000, no error.

**Comment by EtienneWan** [comment]
With Firefox 39, with a payload of 1024, I get this in the console:
```js
message { target: WebSocket, isTrusted: true, data: "941", origin: "wss://squadserver-php-etiennewan-1.c9.io", lastEventId: "", currentTarget: WebSocket, eventPhase: 2, bubbles: false, cancelable: false, defaultPrevented: false, timeStamp: 1438242186765000 } test.html:21:9

close { target: WebSocket, isTrusted: true, wasClean: true, code: 1002, reason: "Get rsv1: 0, rsv2: 1, rsv3: 0, they all must be equal to 0.", currentTarget: WebSocket, eventPhase: 2, bubbles: false, cancelable: false, defaultPrevented: false, timeStamp: 1438242186769000 } test.html:16:9
```
The server echoes 941, than nothing else, as the connection was closed.

**Comment by EtienneWan** [comment]
I saw your test with 1000 and 10000 in the server console, I don't understand why it did work with you ?!?

**Comment by Hywan** [comment]
@EtienneWan It possibly is a bug in Firefox that computes corrupted packets. Test with Chrome, Safari or a newer version of Firefox?

**Comment by Hywan** [comment]
@EtienneWan How long should I wait to get your error (I am testing with Firefox nightly)?

**Comment by EtienneWan** [comment]
For me, it's always the first or second message which is partially received, and then the connection closed.
Tried to test with IE11 with 1024 of payload, get the same error:
```js
[object MessageEvent]
   {
      [functions]: ,
      __proto__: { },
      AT_TARGET: 2,
      bubbles: false,
      BUBBLING_PHASE: 3,
      cancelable: false,
      cancelBubble: false,
      CAPTURING_PHASE: 1,
      constructor: { },
      currentTarget: { },
      data: "935",
      defaultPrevented: false,
      eventPhase: 2,
      isTrusted: true,
      origin: "",
      ports: null,
      source: null,
      srcElement: { },
      target: { },
      timeStamp: 1438243379467,
      type: "message"
   }

[object CloseEvent]
   {
      [functions]: ,
      __proto__: { },
      AT_TARGET: 2,
      bubbles: false,
      BUBBLING_PHASE: 3,
      cancelable: false,
      cancelBubble: false,
      CAPTURING_PHASE: 1,
      code: 1005,
      constructor: { },
      currentTarget: { },
      defaultPrevented: false,
      eventPhase: 2,
      isTrusted: true,
      reason: "Get rsv1: 0, rsv2: 1, rsv3: 0, they all must be equal to 0.",
      srcElement: { },
      target: { },
      timeStamp: 1438243379472,
      type: "close",
      wasClean: true
   }
```

**Comment by EtienneWan** [comment]
Chrome 44: (payload 1024), the server closes connection on first message, received "952"
Safari: (payload 1024), the server closes connection on first message too, received "976"

**Comment by Hywan** [comment]
What about Firefox nightly?

**Comment by osaris** [comment]
No problem with a 1024 payload on these browsers :

* Mac Chrome Version 44.0.2403.89 (64-bit)
* Mac Safari Version 8.0.7 (10600.7.12)
* Windows Chrome Version 44.0.2403.125 m
* Mac Firefox 39.0

**Comment by EtienneWan** [comment]
What can be the source of this error ? It seems not to be relative to the browser, not to be relative to the Internet conection (I'm testing at home and with SauceLabs browsers directly in cloud9, I'm always getting this error, even for the browser you said there were no problem).
This is akward, when I'm the only one who can reproduce as many times as I want.
I can see your tests on my server console without problems, this is definitely mind-blowing

**Comment by Hywan** [comment]
@EtienneWan Any proxy?

**Comment by EtienneWan** [comment]
No proxy, for as long as I know. Any method to check if I got a proxy ?

**Comment by EtienneWan** [comment]
A bit hardcore, but may Wireshark on the client side (maybe server-side too ?) help ?

**Comment by EtienneWan** [comment]
Firefox nightly on Linux (SauceLabs, payload of 800): only one message received ("762"), then connection closed (without CloseEvent logged to the console)

**Comment by Hywan** [comment]
@EtienneWan Did you test with another Internet connection or another computer?

**Comment by EtienneWan** [comment]
Yes, from my desktop and my laptop on the same Internet connection, through SauceLabs (Virtual Machines for cross-browser testing in the cloud)

**Comment by Hywan** [comment]
@EtienneWan Very strange :-/. Did you ask to anybody else around you?

Please @hoaproject/hoackers, can you test https://squadclient-html5-etiennewan-1.c9.io/test.html with `1024` and the console opened. If it does not fail, then it's good. Report results here please.

**Comment by EtienneWan** [comment]
Tested with Chrome 44 on another computer at home a minute ago, it didn't fail with 1024 for a few minutes, but failed on the first message with 4096 (received "3795", then connection closed with no reason in the event)

De:"Ivan Enderlin" <notifications@github.com>
Date:jeu. j juil. PM à 13:27
Objet:Re: [Websocket] "Get rsv1: 0, rsv2: 1, rsv3: 0, they all must be equal to 0" with large arrays sent (#46)

@EtienneWan Very strange :-/. Did you ask to anybody else around you?

Please @hoaproject/hoackers, can you test https://squadclient-html5-etiennewan-1.c9.io/test.html with 1024 and the console opened. If it does not fail, then it's good. Report results here please.

**Comment by Hywan** [comment]
@EtienneWan Works here :-(. @osaris?

**Comment by camael24** [comment]
Hi , i test your page and i give this with an payload 1024 : 
- http://pastebin.com/AM4efQYX => Host : windows 8.1, Firefox 39, at work (no proxy)
- http://pastebin.com/xj1B2V2R => Host : windows 8.1, Chrome 44, at work (no proxy)

Cheers

**Comment by Hywan** [comment]
So with Firefox 39 it does not work. @camael24 Could you try with another version (>39) of Firefox? Same question for @EtienneWan please.

**Comment by EtienneWan** [comment]
With Firefox Developper Edition (41.0a2), payload 1024, only one message received : "977", but there is no close Event logged to the console.

**Comment by Hywan** [comment]
@EtienneWan To clarify, `rsv1`, `rsv2` and `rsv3` are reserved fields for future extension of the WebSocket protocol. They must all be equal to 0 now because no extension exists yet. If we receive a packet with these values not set to 0, this is a bug from the **client**, not from the **server**. I would like to be more confident with the server and its ability to correctly parse a packet than clients to build these packets.
Also, it seems we are not able to reproduce this error. I don't know what to do here. I would have to close this issue but I don't want to let you lost in the wild… What do you expect from us :-/? A solution would be to run wireshark and see if payloads are correctly send. This way we will know which part (client or server) to blame. Thoughts?

**Comment by EtienneWan** [comment]
When I was reading the source code, I did understand the part where the RSV flags are set (server side), but not the part using the ASCII code to set those flags... From that, I understand the incoming packets on the server were influencing the server sent packets, but not if it was client's or the server's fault.

Because you know your library better than me, I believe you the server has no problem. I'll have to test another websocket server (to see if I get the same error, a library with node I think).

I will try to record frames with wireshark (I will do it serverside, it will be easier) and I will post the results here (later today).

I don't hope for anything else, I'm pretty sure my client is too intensive (but why do it fail that way for some people ? It remains odd).

**Comment by Hywan** [comment]
I really want to help you, except if we can find the issue, but I am not able to reproduce and no one is able here (as far as I read the comments of this issue). Please yes, do a wireshark scan on the server side, and maybe, if possible, on the client side too.

**Comment by EtienneWan** [comment]
Firefox 39 capture server side (frames 302 and 309) [https://preview.c9.io/etiennewan/squadserver-php/Firefox39Failed.pcap] (can't upload this file type on github issue)

302: the server says to the client the payload length received (here 514, I'm still sending 1024 of payload)
309: websocket closes connection


Sorry for the bunch of SSH frames

**Comment by Hywan** [comment]
@EtienneWan I am trying something.

**Comment by Hywan** [comment]
@EtienneWan Ok so I locally tried with 65537 (65535) is the last limit for huge message) and it works. I also re-run the test suite from http://autobahn.ws/ and it passed with success.

To sum up. The capture you did with WireShark only shows what the server sent back to the client but not what the server received. This is what I would like to know. Maybe the client only sends a valid packet but I suspect a proxy somewhere to do strange stuff with it.

**Comment by EtienneWan** [comment]
I don't know why wireshark doesn't show the client packets with Websocket level.
Is there any debug options to log all Hoa messages (or packets) to a file ?

**Comment by Hywan** [comment]
@EtienneWan You mean, incoming data through network? No :-(.

**Comment by EtienneWan** [comment]
If wireshark can't display properly client frames, if there is no log options, there is still a way to inspect left.
I can propose you to place some breakpoints on cloud9 and to help me investigate (with Skype/Teamspeak/Firefox Hello). 

If you know why the client packets are not captured at web socket level with wireshark, please tell me :)

**Comment by Hywan** [comment]
@EtienneWan Don't know. We can try a debug session tomorrow at 13h (Europe/Paris), thought?

**Comment by EtienneWan** [comment]
OK for 13h tomorrow :)

**Comment by Hywan** [comment]
Booked.

**Comment by Hywan** [comment]
After a live debugging session, it appears this is an issue with the network. Nor the clients or the server are impacted. The client sends correct data and the server receives corrupted data. So a proxy, in the middle, makes something bad.

Or (I just got this idea), this could be about an encryption. @EtienneWan Is the connection encrypted?

**Comment by Hywan** [comment]
Edit: If the connection were encrypted, I am not sure we were able to do the handshake, so… my proxy idea should be still valid.

**Comment by EtienneWan** [comment]
During the debugging session, we were on HTTP and WS:// in order to live capture with wire shark
We weren't using https and wss

**Comment by EtienneWan** [comment]
Because I didn't set a proxy manually and all tests said there wasn't *most probably* a proxy between me and the server, I'm going to have a talk with my ISP.

I discovered all my computers at home who are using my ISP's DNS servers don't have any problem (aka sending 1024 floats works). My computer I used for the tests is using Google's DNS servers.
Not sure if this is related to my problem, but I got serious doubts right now.

(Out of topic: my ISP is lying about his POP and SMTP servers IP addresses to the other DNS servers such as Google's. That's why I got doubts because I don't use them but it might be not related...)

**Comment by Hywan** [comment]
Thanks.

**Comment by EtienneWan** [comment]
I think you can close this issue, I think no more testing is relevant to bring something new and relevant to this bug.
I'll tell you if I got problems with another library (in this case, it will be my computer's fault), but for the moment, I think that there is nothing more to do.
Closing this issue... :)

**Comment by Hywan** [comment]
Keep us updated please!

**Comment by EtienneWan** [comment]
I don't know if it is needed to reopen this issue, but I got updates.
I'm Using node module ws to make an analog WebSocket server and to do some tests.

```js
var WebSocketServer = require('ws').Server
  , wss = new WebSocketServer({ port: 8080 });

wss.on('connection', function connection(ws) {
  console.log("new client connection");

  ws.on('message', function incoming(message) {
    var received = message.split(',').length;
    console.log(received);
    //ws.send({"size": received});
  });
});

console.log("Server is starting");
```
(note that the commented line caused me problems, such as 'TypeError: Invalid non-string/buffer chunk' thrown in the console)
Without the server sending a response to the client, I can send arrays to the JS websocket implementation... 256 to 65536 floats WITHOUT ANY PROBLEM !!

I tried to comment the line sending a response to the client in Server.php ( ``` $source->send($received)```), it seems that sending 1024 of payload lasts a bit longer (it only fails at the 11th message sent by the client). 4096 of payload fails at first message (as always for me)

No changes were made to the client.
I will try another PHP implementation to see if this issue is PHP oriented or something else...
I'm on a road full of surprise !
